File: src/core/shared/message_types.rs
use serde::{Deserialize, Serialize};

/// Enum representing different types of messages in the bot system
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(transparent)]
pub struct MessageType(pub i32);

impl MessageType {
    /// Regular message from external systems (WhatsApp, Instagram, etc.)
    pub const EXTERNAL: MessageType = MessageType(0);

    /// User message from web interface
    pub const USER: MessageType = MessageType(1);

    /// Bot response (can be regular content or event)
    pub const BOT_RESPONSE: MessageType = MessageType(2);

    /// Continue interrupted response
    pub const CONTINUE: MessageType = MessageType(3);

    /// Suggestion or command message
    pub const SUGGESTION: MessageType = MessageType(4);

    /// Context change notification
    pub const CONTEXT_CHANGE: MessageType = MessageType(5);
}

impl From<i32> for MessageType {
    fn from(value: i32) -> Self {
        MessageType(value)
    }
}

impl From<MessageType> for i32 {
    fn from(value: MessageType) -> Self {
        value.0
    }
}

impl Default for MessageType {
    fn default() -> Self {
        MessageType::USER
    }
}

impl std::fmt::Display for MessageType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let name = match self.0 {
            0 => "EXTERNAL",
            1 => "USER",
            2 => "BOT_RESPONSE",
            3 => "CONTINUE",
            4 => "SUGGESTION",
            5 => "CONTEXT_CHANGE",
            _ => "UNKNOWN",
        };
        write!(f, "{}", name)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_message_type_conversion() {
        assert_eq!(i32::from(MessageType::USER), 1);
        assert_eq!(MessageType::from(2), MessageType::BOT_RESPONSE);
    }

    #[test]
    fn test_message_type_display() {
        assert_eq!(MessageType::USER.to_string(), "USER");
        assert_eq!(MessageType::BOT_RESPONSE.to_string(), "BOT_RESPONSE");
    }

    #[test]
    fn test_message_type_equality() {
        assert_eq!(MessageType::USER, MessageType(1));
        assert_ne!(MessageType::USER, MessageType::BOT_RESPONSE);
    }
}

File: botui/ui/shared/messageTypes.js
/**
 * Message Type Constants
 * Defines the different types of messages in the bot system
 * These values must match the server-side MessageType enum in Rust
 */

const MessageType = {
    /** Regular message from external systems (WhatsApp, Instagram, etc.) */
    EXTERNAL: 0,

    /** User message from web interface */
    USER: 1,

    /** Bot response (can be regular content or event) */
    BOT_RESPONSE: 2,

    /** Continue interrupted response */
    CONTINUE: 3,

    /** Suggestion or command message */
    SUGGESTION: 4,

    /** Context change notification */
    CONTEXT_CHANGE: 5
};

/**
 * Get the name of a message type
 * @param {number} type - The message type number
 * @returns {string} The name of the message type
 */
function getMessageTypeName(type) {
    const names = {
        0: 'EXTERNAL',
        1: 'USER',
        2: 'BOT_RESPONSE',
        3: 'CONTINUE',
        4: 'SUGGESTION',
        5: 'CONTEXT_CHANGE'
    };
    return names[type] || 'UNKNOWN';
}

/**
 * Check if a message is a bot response
 * @param {Object} message - The message object
 * @returns {boolean} True if the message is a bot response
 */
function isBotResponse(message) {
    return message && message.message_type === MessageType.BOT_RESPONSE;
}

/**
 * Check if a message is a user message
 * @param {Object} message - The message object
 * @returns {boolean} True if the message is from a user
 */
function isUserMessage(message) {
    return message && message.message_type === MessageType.USER;
}

/**
 * Check if a message is a context change
 * @param {Object} message - The message object
 * @returns {boolean} True if the message is a context change
 */
function isContextChange(message) {
    return message && message.message_type === MessageType.CONTEXT_CHANGE;
}

/**
 * Check if a message is a suggestion
 * @param {Object} message - The message object
 * @returns {boolean} True if the message is a suggestion
 */
function isSuggestion(message) {
    return message && message.message_type === MessageType.SUGGESTION;
}

// Export for use in other modules (if using modules)
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        MessageType,
        getMessageTypeName,
        isBotResponse,
        isUserMessage,
        isContextChange,
        isSuggestion
    };
}

// Also make available globally for non-module scripts
if (typeof window !== 'undefined') {
    window.MessageType = MessageType;
    window.getMessageTypeName = getMessageTypeName;
    window.isBotResponse = isBotResponse;
    window.isUserMessage = isUserMessage;
    window.isContextChange = isContextChange;
    window.isSuggestion = isSuggestion;
}
